include/AdePT/core/HostScoringImpl.cuh

```diff
@@ -98,145 +106,147 @@ __device__ __forceinline__ void refresh_stats(HostScoring *hostScoring_dev)
+{
+  // Increment number of secondaries
+  atomicAdd(&hostScoring_dev->fGlobalCounters_dev->numElectrons, num_ele);
+  atomicAdd(&hostScoring_dev->fGlobalCounters_dev->numPositrons, num_pos);
+  atomicAdd(&hostScoring_dev->fGlobalCounters_dev->numGammas, num_gam);
+}
 
-  template <typename IntegrationLayer>
-  inline void EndOfIteration(HostScoring &hostScoring, HostScoring *hostScoring_dev, cudaStream_t &stream, IntegrationLayer &integration)
-  {
-    // Copy host scoring stats from device to host
-    // COPCORE_CUDA_CHECK(
-    //       cudaMemcpyAsync(&hostScoring.fStats, hostScoring.fStats_dev, sizeof(HostScoring::Stats), cudaMemcpyDeviceToHost, stream));
-    // COPCORE_CUDA_CHECK(cudaStreamSynchronize(stream));
-
-    // Check if we need to flush the hits buffer
-    if (CheckAndFlush(hostScoring, hostScoring.fStats, hostScoring_dev, stream)) {
-      // Synchronize the stream used to copy back the hits
-      COPCORE_CUDA_CHECK(cudaStreamSynchronize(stream));
-      // Process the hits on CPU
-      for (const auto &hit : hostScoring) {
-        integration.ProcessGPUHit(hit);
-      }
-    }
-  }
+template <>
+__device__ __forceinline__ void EndOfIterationGPU(HostScoring *hostScoring_dev)
+{
+  // Update hit buffer stats
+  refresh_stats(hostScoring_dev);
+}
 
-  template <typename IntegrationLayer>
-  inline void EndOfTransport(HostScoring &hostScoring, HostScoring *hostScoring_dev, cudaStream_t &stream, IntegrationLayer &integration)
-  {
-    // Transfer back scoring.
-    CopyHitsToHost(hostScoring, hostScoring.fStats, hostScoring_dev, stream);
-    // Transfer back the global counters
-    // scoring->fGlobalCounters_dev->numKilled = inFlight;
-    CopyGlobalCountersToHost(hostScoring, stream);
+template <typename IntegrationLayer>
+inline void EndOfIteration(HostScoring &hostScoring, HostScoring *hostScoring_dev, cudaStream_t &stream,
+                           IntegrationLayer &integration)
+{
+  // Copy host scoring stats from device to host
+  // COPCORE_CUDA_CHECK(
+  //       cudaMemcpyAsync(&hostScoring.fStats, hostScoring.fStats_dev, sizeof(HostScoring::Stats),
+  //       cudaMemcpyDeviceToHost, stream));
+  // COPCORE_CUDA_CHECK(cudaStreamSynchronize(stream));
+
+  // Check if we need to flush the hits buffer
+  if (CheckAndFlush(hostScoring, hostScoring_dev, stream)) {
+    // Synchronize the stream used to copy back the hits
     COPCORE_CUDA_CHECK(cudaStreamSynchronize(stream));
-    // Process the last hits on CPU
+    // Process the hits on CPU
     for (const auto &hit : hostScoring) {
       integration.ProcessGPUHit(hit);
     }
   }
 }
 
-
-
+template <typename IntegrationLayer>
+inline void EndOfTransport(HostScoring &hostScoring, HostScoring *hostScoring_dev, cudaStream_t &stream,
+                           IntegrationLayer &integration)
+{
+  // Transfer back scoring.
+  CopyHitsToHost(hostScoring, hostScoring_dev, stream);
+  ResetBufferStatsGPU<<<1, 1, 0, stream>>>(hostScoring_dev);
```

Comment:
Why launch a new kernel to reset this and not call a Reset function inside say `UpdateBufferStartGPU` where we already pass the hostScoring_dev pointer?